py3status/modules/google_calendar.py

```diff
@@ -0,0 +1,502 @@
+
+        Returns: Credentials, the obtained credential.
+        """
+        client_secret_path = os.path.dirname(self.client_secret)
+        auth_token_path = os.path.dirname(self.auth_token)
+
+        if not os.path.exists(auth_token_path):
+            os.makedirs(auth_token_path)
+
+        if not os.path.exists(client_secret_path):
+            os.makedirs(client_secret_path)
+
+        flags = tools.argparser.parse_args(args=[])
+        store = Storage(self.auth_token)
+        credentials = store.get()
+
+        if not credentials or credentials.invalid:
+            try:
+                flow = client.flow_from_clientsecrets(self.client_secret, SCOPES)
+                flow.user_agent = APPLICATION_NAME
+                if flags:
+                    credentials = tools.run_flow(flow, store, flags)
+                else:  # Needed only for compatibility with Python 2.6
+                    credentials = tools.run(flow, store)
+            except clientsecrets.InvalidClientSecretsError:
+                raise Exception('missing client_secret')
+
+            """
+            Have to restart i3 after getting credentials to prevent bad output.
+            This only has to be done once on the first run of the module.
+            """
+            self.py3.command_run('i3-msg restart')
+
+        return credentials
+
+    def _authorize_credentials(self):
+        """
+        Fetches an access/refresh token by authorizing OAuth 2.0 credentials.
+
+        Returns: True, if the authorization was successful.
+                 False, if a ServerNotFoundError is thrown.
+        """
+        try:
+            http = self.credentials.authorize(httplib2.Http())
+            self.service = discovery.build('calendar', 'v3', http=http)
+            return True
+        except ServerNotFoundError:
+            return False
+
+    def _get_events(self):
+        """
+        Fetches events from the color into a list.
+
+        Returns: The list of events.
+        """
+        self.last_update = datetime.datetime.now()
+        now = datetime.datetime.utcnow().isoformat() + 'Z'  # 'Z' indicates UTC time
+
+        event_cache = self.events
+
+        try:
+            eventsResult = self.service.events().list(
+                calendarId='primary', timeMin=now, maxResults=self.num_events, singleEvents=True,
+                orderBy='startTime').execute(num_retries=5)
+        except:
+            return event_cache
+
+        return eventsResult.get('items', [])
+
+    def _check_warn_threshold(self, time_to, start, end, summary):
+        """
+        Checks if the time until an event starts is less than or equal to the warn_threshold.
+        If it is, it presents a warning with self.py3.notify_user.
+        """
+        if time_to['total_minutes'] <= self.warn_threshold:
+            warn_message = str(summary) + " " + str(start) + " - " + str(end)
+            self.py3.notify_user(warn_message, 'warning', self.warn_timeout)
```

Comment:
(Q) Should we make `format_notification` here?

Reply:
Done.

Reply:
I think if you do this... `event_dict['format_timer'] = self._format_timedelta(index, time_delta, is_current)`, then you could just pass in the dict ... in `_check_warn_thresholds`. Skip unnecessary specific lines + names.

Reply:
Oh damn, that was silly of me. Good catch.